src/api/libcellml/issue.h

```diff
@@ -0,0 +1,433 @@
+     *
+     * Set the @p cause of this issue from the options available in
+     * @c Issue::Cause.
+     *
+     * @param cause The @c Issue::Cause to set.
+     */
+    void setCause(Cause cause);
+
+    /**
+     * @brief Get the cause of this issue.
+     *
+     * Get the @c cause of this issue. If no cause has been set for
+     * this issue, will return Cause::UNDEFINED.
+     *
+     * @return The @c Issue::Cause set for this issue.
+     */
+    Cause cause() const;
+
+    /**
+     * @brief Check whether the cause of this issue matches the argument cause.
+     *
+     * Return @c true if the @p cause matches the the @c Issue::Cause for this
+     * issue and @c false otherwise.
+     *
+     * @return @c true if the @p cause matches the the @c Issue::Cause for this
+     * issue and @c false otherwise.
+     */
+    bool isCause(Cause cause) const;
+
+    /**
+     * @brief Set the level of this issue.
+     *
+     * Set the @p level of this issue from the options available in
+     * @c Issue::Level.
+     *
+     * @param level The @c Issue::Level to set.
+     */
+    void setLevel(Level level);
+
+    /**
+     * @brief Get the level of this issue.
+     *
+     * Get the @c level of this issue. If no level has been set for
+     * this issue, will return Level::ERROR.
+     *
+     * @return The @c Issue::Level set for this issue.
+     */
+    Level level() const;
+
+    /**
+     * @brief Check whether the level of this issue matches the argument level.
+     *
+     * Return @c true if the @p level matches the the @c Issue::Level for this
+     * issue and @c false otherwise.
+     *
+     * @return @c true if the @p level matches the the @c Issue::Level for this
+     * issue and @c false otherwise.
+     */
+    bool isLevel(Level level) const;
+
+    /**
+     * @brief Set the @c enum ReferenceRule of this issue.
+     *
+     * Set the @p rule of this issue from the options available in
+     * the @c ReferenceRule @c enum.
+     *
+     * @param rule The @c ReferenceRule to set.
+     */
+    void setRule(ReferenceRule rule);
+
+    /**
+     * @brief Get the @c enum ReferenceRule of this issue.
+     *
+     * Get the @c enum @c ReferenceRule value @p rule of this issue. If
+     * no rule has been set for this issue, will return ReferenceRule::UNDEFINED.
+     *
+     * @return The @c ReferenceRule for this issue.
+     */
+    ReferenceRule rule() const;
+
+    /**
+     * @brief Set the @c url of this issue.
+     *
+     * Manually set the @c url to which the user should refer for more information on the
+     * specification rule which has been broken or guidelines that haven't been followed.
+     * If this has not been set manually by use of this function, the default root url
+     * will be used (specified in baseIssueUrl) in conjunction with the referenceHeading
+     * string.
+     *
+     * @param rule The @c url to set manually.
+     */
+    void setUrl(std::string &url) const;
```

Comment:
this seems problematic to me. The `ReferenceRule`'s should be all you need, they should each have a corresponding URL. I'd be concerned that you could start getting user-code setting URLs that look in errors as if they come from libCellML but they do not.

Reply:
Well spotted and fully agreed.

Reply:
It was included for two reasons: 
- situations where we want to refer to external sources (eg, the MathML guidelines or other namespacey stuff).  The alternative is to have a chapter in the references which lists all the external links, so the user has to go through our site to get to the external one.
- situations where you just want to give more information, but it's not tied to a reference.  Since we have a `HINT` level, this could include a url pointing to the tutorials, or something in PMR, or whatever.  Haven't quite figured out how best to use it yet though ...

Reply:
(but you're right, this function is only there for the outlying cases, the others will come directly from the rule number)

Reply:
I've removed it now.  I suggest we add a chapter to the References (= Spec + everything else) that deals with external sites and namespaces.